Sources/AsyncKit/ConnectionPool.swift

```diff
@@ -216,23 +249,41 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
     /// Connection pools must be closed before they deinitialize.
     ///
     /// - returns: A future indicating close completion.
-    public func close() -> EventLoopFuture<Void> {
-        self.isClosed = true
-        return self.available.map { $0.close() }.flatten(on: self.eventLoop).map {
-            // inform any waiters that they will never be receiving a connection
-            while let waiter = self.waiters.popFirst() {
-                waiter.fail(ConnectionPoolError.closed)
+    public func shutdown() {
+        // synchronize access to closing
+        self.lock.lock()
+
+        // check to make sure we aren't double closing
+        guard !self.didShutdown else {
+            return
+        }
+        self.didShutdown = true
+
+        // is closed variable mutated
+        self.lock.unlock()
+
+        // no locks needed as this can only happen once
+        for available in self.available {
+            do {
```

Comment:
What about closing connections that are currently in use? Are those kept track of by this class at all? (I couldn’t find where)

Reply:
If the `releaseConnection` is called after the pool is shutdown, it is closed instead of added to the available array.

Reply:
I wonder if we should close _all_ connections _immediately_ when the pool gets shut down, though, even the ones that are in use. Otherwise, we can't guarantee how long connections will stay open if clients take long to release them. WDYT?

Reply:
Yeah the current design assumes that all shut downs will be graceful and the client should eventually give back the connection. I _think_ this is a fair assumption. It seems acceptable to me that a connection pool might receive a shutdown signal while an HTTP request is still in flight and a connection is being used. Assuming the HTTP server shuts down gracefully (which it does in v4) then its totally fine for the request to continue to completion even after shut down is signaled. 
However, this is all making a lot of assumptions. Maybe we want to keep a record of all active connections and ensure they are returned before the pool deinits. We could do this in debug builds only to help track down bugs without hurting release performance. 
FWIW, we already have protection against this at the connection level. Psql, mysql, and sqlite connections will all assert if they deinit before being closed.

Reply:
Hmm, to me shutting down a connection pool also means closing the attached connections. Maybe add an option to close all connections to the `shutdown` method?